File: scripts/sec-insider-buys.mjs
/**
 * SEC EDGAR Insider Purchase Scanner
 *
 * Fetches Form 4 filings from the last 24 hours, filters for executive/director
 * cash purchases >$100k, ranks by value, and posts a summary to Slack.
 *
 * Required env vars (one of the two Slack options):
 *   SLACK_WEBHOOK_URL             — Incoming Webhook URL
 *   SLACK_BOT_TOKEN + SLACK_CHANNEL_ID  — Bot token + channel
 *
 * Optional:
 *   SEC_USER_AGENT   — "CompanyName email@example.com" (SEC requires identification)
 *   MIN_PURCHASE_USD — Minimum purchase value in USD (default: 100000)
 *   DRY_RUN          — Set to "1" to print the message instead of sending to Slack
 */

const EDGAR_SEARCH_URL = 'https://efts.sec.gov/LATEST/search-index';
const EDGAR_ARCHIVES_URL = 'https://www.sec.gov/Archives/edgar/data';
const USER_AGENT = process.env.SEC_USER_AGENT || 'TradingJournalPro research@tradingjournalpro.com';
const MIN_PURCHASE_USD = Number(process.env.MIN_PURCHASE_USD) || 100_000;
const DRY_RUN = process.env.DRY_RUN === '1';

// SEC rate limit: 10 req/s — we stay well under at ~6 req/s
const RATE_LIMIT_MS = 160;

// ─── Utilities ───────────────────────────────────────────────────────────────

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

function formatDate(date) {
  return date.toISOString().split('T')[0];
}

function formatCurrency(amount) {
  if (amount >= 1_000_000) return `$${(amount / 1_000_000).toFixed(2)}M`;
  if (amount >= 1_000) return `$${(amount / 1_000).toFixed(0)}K`;
  return `$${amount.toFixed(0)}`;
}

/** Extract all text between <tag>…</tag> occurrences as an array of strings. */
function extractBlocks(xml, tag) {
  const blocks = [];
  const startTag = `<${tag}>`;
  const endTag = `</${tag}>`;
  let pos = 0;
  while (true) {
    const start = xml.indexOf(startTag, pos);
    if (start === -1) break;
    const end = xml.indexOf(endTag, start + startTag.length);
    if (end === -1) break;
    blocks.push(xml.slice(start + startTag.length, end));
    pos = end + endTag.length;
  }
  return blocks;
}

/**
 * Extract the text value of a tag from a block.
 * Handles both <tag>direct</tag> and <tag><value>nested</value></tag>.
 */
function extractValue(block, tag) {
  const inner = extractBlocks(block, tag)[0];
  if (!inner) return null;
  const valueBlock = extractBlocks(inner, 'value')[0];
  return (valueBlock ?? inner).trim();
}

// ─── SEC EDGAR: fetch filings list ───────────────────────────────────────────

async function fetchWithRetry(url, retries = 3) {
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      const res = await fetch(url, { headers: { 'User-Agent': USER_AGENT } });
      if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
      return res;
    } catch (err) {
      if (attempt === retries) throw err;
      await sleep(1000 * 2 ** attempt);
    }
  }
}

/**
 * Fetch all Form 4 filings filed in the given date range from EDGAR full-text search.
 * Returns the raw search hits array.
 */
async function fetchForm4Filings(startDate, endDate) {
  const start = formatDate(startDate);
  const end = formatDate(endDate);
  const allHits = [];
  let from = 0;
  const size = 40;
  let total = Infinity;

  while (from < total) {
    const url = `${EDGAR_SEARCH_URL}?q=&forms=4&dateRange=custom&startdt=${start}&enddt=${end}&from=${from}&size=${size}`;
    try {
      const res = await fetchWithRetry(url);
      const data = await res.json();
      total = data.hits?.total?.value ?? 0;
      const hits = data.hits?.hits ?? [];
      allHits.push(...hits);
      if (hits.length < size) break;
      from += size;
      await sleep(RATE_LIMIT_MS);
    } catch (err) {
      console.error(`EDGAR search page from=${from} failed: ${err.message}`);
      break;
    }
  }

  return allHits;
}

// ─── SEC EDGAR: resolve XML URL for a filing ─────────────────────────────────

/**
 * From the accession number, try to derive the filer CIK.
 * Accession format: 0001234567-YY-NNNNNN — first segment is often the filer CIK.
 * Used as a fallback when entity_id is absent.
 */
function cikFromAccession(accessionNo) {
  return accessionNo.split('-')[0] ?? null;
}

/**
 * Fetch the filing index JSON and return the URL of the primary Form 4 XML.
 */
async function resolveXmlUrl(cik, accessionNo) {
  const accFormatted = accessionNo.replace(/-/g, '');
  const indexUrl = `${EDGAR_ARCHIVES_URL}/${cik}/${accFormatted}/${accessionNo}-index.json`;

  try {
    const res = await fetchWithRetry(indexUrl);
    const index = await res.json();
    const docs = index.documents ?? [];

    // Prefer the document explicitly typed "4"
    const primary = docs.find(d => d.type === '4' && d.document?.endsWith('.xml'))
      ?? docs.find(d => d.document?.endsWith('.xml'));

    if (primary) {
      return `${EDGAR_ARCHIVES_URL}/${cik}/${accFormatted}/${primary.document}`;
    }
  } catch {
    // Ignore index fetch failures — the caller will skip this filing
  }
  return null;
}

// ─── Form 4 XML parser ───────────────────────────────────────────────────────

/**
 * Parse a Form 4 XML string and return an array of purchase records.
 * Only returns non-derivative cash purchases (transaction code "P").
 */
function parseForm4(xml, entityName, displayNames, fileDate) {
  const issuerName = extractValue(xml, 'issuerName') || entityName || 'Unknown';
  const ticker = extractValue(xml, 'issuerTradingSymbol') || '';

  // Owner identity — use first reporting owner
  const ownerBlock = extractBlocks(xml, 'reportingOwner')[0] ?? xml;
  const ownerName =
    extractValue(ownerBlock, 'rptOwnerName') ||
    (displayNames?.[0] ?? '').split(' (')[0] ||
    'Unknown';

  const relBlock = extractBlocks(ownerBlock, 'reportingOwnerRelationship')[0] ?? '';
  const isDirector = extractValue(relBlock, 'isDirector') === '1';
  const isOfficer = extractValue(relBlock, 'isOfficer') === '1';
  const officerTitle = extractValue(relBlock, 'officerTitle') || '';
  const role = isOfficer && officerTitle
    ? officerTitle
    : isDirector ? 'Director'
    : isOfficer ? 'Officer'
    : 'Insider';

  const purchases = [];

  for (const txnBlock of extractBlocks(xml, 'nonDerivativeTransaction')) {
    const code = extractValue(txnBlock, 'transactionCode');
    const acquired = extractValue(txnBlock, 'transactionAcquiredDisposedCode');

    if (code !== 'P' || acquired !== 'A') continue;

    const sharesStr = extractValue(txnBlock, 'transactionShares');
    const priceStr = extractValue(txnBlock, 'transactionPricePerShare');

    const shares = parseFloat(sharesStr ?? '');
    const price = parseFloat(priceStr ?? '');

    if (!isFinite(shares) || !isFinite(price) || price <= 0 || shares <= 0) continue;

    purchases.push({
      issuerName,
      ticker: ticker.trim(),
      ownerName: ownerName.trim(),
      role,
      shares,
      price,
      totalValue: shares * price,
      securityTitle: extractValue(txnBlock, 'securityTitle') || 'Common Stock',
      transactionDate: extractValue(txnBlock, 'transactionDate') || fileDate,
      fileDate,
    });
  }

  return purchases;
}

// ─── Main pipeline ───────────────────────────────────────────────────────────

async function processFilings(filings) {
  const allPurchases = [];
  const total = filings.length;

  for (let i = 0; i < total; i++) {
    const { _source: src } = filings[i];
    const accessionNo = src?.accession_no;
    // entity_id is the filer CIK when present; fall back to accession prefix
    const cik = src?.entity_id || cikFromAccession(accessionNo ?? '');

    if (!accessionNo || !cik) continue;

    try {
      await sleep(RATE_LIMIT_MS);
      const xmlUrl = await resolveXmlUrl(cik, accessionNo);
      if (!xmlUrl) continue;

      await sleep(RATE_LIMIT_MS);
      const xmlRes = await fetchWithRetry(xmlUrl);
      const xml = await xmlRes.text();

      const purchases = parseForm4(
        xml,
        src.entity_name ?? '',
        src.display_names ?? [],
        (src.file_date ?? '').split('T')[0],
      );
      allPurchases.push(...purchases);
    } catch {
      // Skip failed filings silently — they're rare but expected
    }

    if ((i + 1) % 50 === 0 || i + 1 === total) {
      process.stdout.write(`\r  Processed ${i + 1}/${total} filings — ${allPurchases.length} purchases found`);
    }
  }

  if (total > 0) console.log(); // newline after progress
  return allPurchases;
}

/**
 * Aggregate purchases by (issuerName, ownerName) so multiple transactions
 * from the same person on the same day are combined into one entry.
 */
function aggregatePurchases(purchases) {
  const map = new Map();

  for (const p of purchases) {
    const key = `${p.issuerName}::${p.ownerName}`;
    if (map.has(key)) {
      const existing = map.get(key);
      existing.totalValue += p.totalValue;
      existing.shares += p.shares;
      // Keep the latest price (approximate — weighted avg would require more data)
      existing.price = p.price;
    } else {
      map.set(key, { ...p });
    }
  }

  return [...map.values()];
}

// ─── Slack notification ───────────────────────────────────────────────────────

function buildSlackText(purchases, dateRange) {
  const header = `📊 *SEC Insider Purchases — Last 24 Hours*\n_${dateRange} · Cash purchases >$${(MIN_PURCHASE_USD / 1000).toFixed(0)}K by executives & directors_`;

  if (purchases.length === 0) {
    return `${header}\n\n_No qualifying insider purchases found for this period._`;
  }

  const top = purchases.slice(0, 20);
  const rows = top.map((p, i) => {
    const tickerStr = p.ticker ? ` *(${p.ticker})*` : '';
    return [
      `*${i + 1}.* ${p.ownerName} — ${p.role}`,
      `   🏢 ${p.issuerName}${tickerStr}`,
      `   💰 ${p.shares.toLocaleString(undefined, { maximumFractionDigits: 0 })} shares @ $${p.price.toFixed(2)} = *${formatCurrency(p.totalValue)}*`,
      `   📅 Txn: ${p.transactionDate}  |  Filed: ${p.fileDate}`,
    ].join('\n');
  });

  const footer = purchases.length > 20
    ? `_Showing top 20 of ${purchases.length} qualifying purchases, ranked by value._`
    : `_Total: ${purchases.length} insider purchase${purchases.length === 1 ? '' : 's'} above $${(MIN_PURCHASE_USD / 1000).toFixed(0)}K_`;

  return [header, '', ...rows, '', footer].join('\n');
}

async function sendToSlack(text) {
  if (DRY_RUN) {
    console.log('\n─── Slack message (DRY RUN) ───────────────────────────\n');
    console.log(text);
    console.log('\n───────────────────────────────────────────────────────\n');
    return;
  }

  const webhookUrl = process.env.SLACK_WEBHOOK_URL;
  const botToken = process.env.SLACK_BOT_TOKEN;
  const channelId = process.env.SLACK_CHANNEL_ID;

  if (!webhookUrl && !(botToken && channelId)) {
    console.warn('[warn] No Slack credentials set. Printing message instead.\n');
    console.log(text);
    console.warn(
      '\nSet SLACK_WEBHOOK_URL  —or—  SLACK_BOT_TOKEN + SLACK_CHANNEL_ID to send to Slack.',
    );
    return;
  }

  if (webhookUrl) {
    const res = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text, mrkdwn: true }),
    });
    if (!res.ok) throw new Error(`Slack webhook failed: ${res.status} ${await res.text()}`);
    console.log('✓ Message sent via Slack Incoming Webhook');
  } else {
    const res = await fetch('https://slack.com/api/chat.postMessage', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${botToken}`,
      },
      body: JSON.stringify({ channel: channelId, text, mrkdwn: true }),
    });
    const data = await res.json();
    if (!data.ok) throw new Error(`Slack API error: ${data.error}`);
    console.log(`✓ Message sent to Slack channel ${channelId}`);
  }
}

// ─── Entry point ─────────────────────────────────────────────────────────────

async function main() {
  const now = new Date();
  const since = new Date(now.getTime() - 24 * 60 * 60 * 1000);
  const dateRange = `${formatDate(since)} → ${formatDate(now)}`;

  console.log(`\nSEC EDGAR Insider Purchase Scanner`);
  console.log(`Period : ${dateRange}`);
  console.log(`Filter : cash purchases > $${MIN_PURCHASE_USD.toLocaleString()}`);
  if (DRY_RUN) console.log(`Mode   : DRY RUN (Slack disabled)\n`);

  console.log('\n[1/3] Fetching Form 4 filings from EDGAR…');
  const filings = await fetchForm4Filings(since, now);
  console.log(`  Found ${filings.length} Form 4 filings`);

  if (filings.length === 0) {
    const text = buildSlackText([], dateRange);
    await sendToSlack(text);
    return;
  }

  console.log('\n[2/3] Parsing filings for insider purchases…');
  const rawPurchases = await processFilings(filings);

  // Aggregate, filter, rank
  const purchases = aggregatePurchases(rawPurchases)
    .filter(p => p.totalValue >= MIN_PURCHASE_USD)
    .sort((a, b) => b.totalValue - a.totalValue);

  console.log(`  Qualifying purchases (>${formatCurrency(MIN_PURCHASE_USD)}): ${purchases.length}`);

  if (purchases.length > 0) {
    console.log('\n  Top 5 preview:');
    purchases.slice(0, 5).forEach((p, i) => {
      const ticker = p.ticker ? ` (${p.ticker})` : '';
      console.log(`    ${i + 1}. ${p.ownerName} — ${p.issuerName}${ticker} — ${formatCurrency(p.totalValue)}`);
    });
  }

  console.log('\n[3/3] Sending summary to Slack…');
  const text = buildSlackText(purchases, dateRange);
  await sendToSlack(text);
}

main().catch(err => {
  console.error('\nFatal error:', err.message);
  process.exit(1);
});

File: scripts/sec_insider_buys.py
#!/usr/bin/env python3
"""
SEC EDGAR Insider Buying Scanner

Fetches Form 4 filings from the last 24 hours, filters for executive/director
open-market purchases > $100k, and returns a ranked list by dollar value.

Usage:
    python3 scripts/sec_insider_buys.py

Requirements:
    pip install requests

NOTE: SEC EDGAR (www.sec.gov / data.sec.gov / efts.sec.gov) blocks requests
from cloud-provider IP ranges (AWS, GCP, Azure). Run this script from a
residential or corporate network, or use a proxy/VPN that routes through a
non-cloud IP. The SEC fair-access policy also requires a descriptive User-Agent
header (set via HEADERS below). See https://www.sec.gov/os/accessing-edgar-data
"""

import requests
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta
import json
import time
import sys

# SEC requires an identifying User-Agent per their fair-access policy
HEADERS = {
    "User-Agent": "TradingJournalPro WILFRED.GUTIERREZ@gmail.com",
    "Accept-Encoding": "gzip, deflate",
    "Accept": "application/json, text/html, */*",
}

MIN_PURCHASE_VALUE = 100_000   # $100k threshold
MAX_FILINGS = 80               # cap to stay well within rate limits
REQUEST_DELAY = 0.12           # 12 req/s max per SEC policy


# ---------------------------------------------------------------------------
# EDGAR helpers
# ---------------------------------------------------------------------------

def get_date_range(days: int = 1):
    now = datetime.now()
    end = now.strftime("%Y-%m-%d")
    start = (now - timedelta(days=days)).strftime("%Y-%m-%d")
    return start, end


def efts_search(start_date: str, end_date: str, size: int = MAX_FILINGS) -> list:
    """
    Call the EDGAR full-text search index for Form 4 filings.
    Returns a list of hit dicts from the response.
    """
    url = "https://efts.sec.gov/LATEST/search-index"
    params = {
        "forms": "4",
        "dateRange": "custom",
        "startdt": start_date,
        "enddt": end_date,
        "hits.hits.total.value": "true",
    }
    resp = requests.get(url, params=params, headers=HEADERS, timeout=30)
    resp.raise_for_status()
    data = resp.json()
    return data.get("hits", {}).get("hits", [])


def efts_search_recent(size: int = MAX_FILINGS) -> list:
    """Fallback: grab the most recent Form 4 filings (no date filter)."""
    url = "https://efts.sec.gov/LATEST/search-index"
    params = {"forms": "4"}
    resp = requests.get(url, params=params, headers=HEADERS, timeout=30)
    resp.raise_for_status()
    data = resp.json()
    return data.get("hits", {}).get("hits", [])[:size]


def get_filing_index(cik: str, acc_dashes: str) -> dict | None:
    """Fetch the filing's index JSON to enumerate documents."""
    acc_nodash = acc_dashes.replace("-", "")
    url = (
        f"https://www.sec.gov/Archives/edgar/data/{cik}/{acc_nodash}/"
        f"{acc_dashes}-index.json"
    )
    try:
        resp = requests.get(url, headers=HEADERS, timeout=20)
        if resp.status_code == 200:
            return resp.json()
    except Exception:
        pass
    return None


def fetch_form4_xml(cik: str, acc_dashes: str, filename: str) -> str | None:
    """Download the Form 4 XML document."""
    acc_nodash = acc_dashes.replace("-", "")
    url = f"https://www.sec.gov/Archives/edgar/data/{cik}/{acc_nodash}/{filename}"
    try:
        resp = requests.get(url, headers=HEADERS, timeout=20)
        if resp.status_code == 200 and resp.text.strip():
            return resp.text
    except Exception:
        pass
    return None


# ---------------------------------------------------------------------------
# XML parser
# ---------------------------------------------------------------------------

def _text(el) -> str:
    return el.text.strip() if el is not None and el.text else ""


def parse_form4(xml_content: str, acc_no: str, file_date: str) -> list:
    """
    Parse a Form 4 XML document and return qualifying purchase records.
    Transaction code 'P' = open-market purchase.
    """
    results = []
    try:
        # Strip namespace declarations that trip up ElementTree
        xml_clean = xml_content.replace(" xmlns=", " _xmlns=")
        root = ET.fromstring(xml_clean)
    except ET.ParseError:
        return results

    # Issuer (company)
    issuer = root.find(".//issuer")
    company = _text(issuer.find("issuerName")) if issuer is not None else ""
    ticker = _text(issuer.find("issuerTradingSymbol")).upper() if issuer is not None else ""

    # Reporting owner (insider)
    owner_name, owner_title = "", ""
    is_director = is_officer = False
    ro = root.find(".//reportingOwner")
    if ro is not None:
        oid = ro.find("reportingOwnerId")
        if oid is not None:
            owner_name = _text(oid.find("rptOwnerName"))
        rel = ro.find("reportingOwnerRelationship")
        if rel is not None:
            is_director = _text(rel.find("isDirector")) == "1"
            is_officer = _text(rel.find("isOfficer")) == "1"
            is_10pct = _text(rel.find("isTenPercentOwner")) == "1"
            owner_title = _text(rel.find("officerTitle"))
            if not owner_title:
                owner_title = "Director" if is_director else ("10% Owner" if is_10pct else "Insider")

    # Non-derivative (common stock) transactions
    for tx in root.findall(".//nonDerivativeTransaction"):
        code = _text(tx.find(".//transactionCode"))
        if code != "P":
            continue

        acq = _text(tx.find(".//transactionAcquiredDisposedCode/value"))
        if acq and acq != "A":
            continue  # skip disposals mis-coded P

        try:
            shares = float(_text(tx.find(".//transactionShares/value")))
            price = float(_text(tx.find(".//transactionPricePerShare/value")))
        except (ValueError, TypeError):
            continue

        if shares <= 0 or price <= 0:
            continue

        total = shares * price
        if total < MIN_PURCHASE_VALUE:
            continue

        tx_date = _text(tx.find(".//transactionDate/value")) or file_date

        results.append(
            {
                "company": company,
                "ticker": ticker,
                "insider": owner_name,
                "title": owner_title,
                "shares": int(shares),
                "price": price,
                "total_value": total,
                "tx_date": tx_date,
                "file_date": file_date,
                "is_director": is_director,
                "is_officer": is_officer,
                "sec_url": f"https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&filenum=&State=0&SIC=&dateb=&owner=include&count=5&search_text=&action=getcompany",
                "filing_url": f"https://www.sec.gov/Archives/edgar/data/",
            }
        )

    return results


# ---------------------------------------------------------------------------
# Orchestrator
# ---------------------------------------------------------------------------

def run_scan(verbose: bool = True) -> list:
    start_date, end_date = get_date_range(days=1)
    log = print if verbose else lambda *a, **k: None

    log(f"[EDGAR scan] date range: {start_date} → {end_date}", flush=True)

    # 1. Get filing list
    hits = []
    try:
        hits = efts_search(start_date, end_date)
        log(f"[EDGAR scan] EFTS returned {len(hits)} Form 4 filings", flush=True)
    except Exception as exc:
        log(f"[EDGAR scan] EFTS error: {exc}", flush=True)

    if not hits:
        log("[EDGAR scan] No dated results – falling back to most-recent feed", flush=True)
        try:
            hits = efts_search_recent()
            log(f"[EDGAR scan] Recent feed: {len(hits)} filings", flush=True)
        except Exception as exc:
            log(f"[EDGAR scan] Recent feed error: {exc}", flush=True)
            return []

    # 2. Process each filing
    all_purchases = []
    processed = 0

    for hit in hits[:MAX_FILINGS]:
        src = hit.get("_source", {})
        file_date = src.get("file_date", end_date)
        acc_dashes = hit.get("_id", "")
        if not acc_dashes or "-" not in acc_dashes:
            continue

        # CIK = first segment of accession number, strip leading zeros
        raw_cik = acc_dashes.split("-")[0].lstrip("0") or "0"

        index = get_filing_index(raw_cik, acc_dashes)
        time.sleep(REQUEST_DELAY)

        if not index:
            continue

        # Find the primary Form 4 XML document
        xml_filename = None
        for doc in index.get("documents", []):
            dtype = doc.get("type", "").strip()
            url_path = doc.get("documentUrl", "") or doc.get("url", "")
            fname = url_path.split("/")[-1] if url_path else ""
            if dtype in ("4", "4/A") and fname.lower().endswith(".xml"):
                xml_filename = fname
                break
        if not xml_filename:
            # Fallback: first XML in the filing
            for doc in index.get("documents", []):
                url_path = doc.get("documentUrl", "") or doc.get("url", "")
                fname = url_path.split("/")[-1] if url_path else ""
                if fname.lower().endswith(".xml"):
                    xml_filename = fname
                    break

        if not xml_filename:
            continue

        xml = fetch_form4_xml(raw_cik, acc_dashes, xml_filename)
        time.sleep(REQUEST_DELAY)

        if not xml:
            continue

        purchases = parse_form4(xml, acc_dashes, file_date)
        all_purchases.extend(purchases)
        processed += 1

        if verbose and purchases:
            for p in purchases:
                log(
                    f"  ✓ {p['ticker'] or p['company']}: {p['insider']} "
                    f"bought {p['shares']:,} sh @ ${p['price']:.2f} "
                    f"= ${p['total_value']:,.0f}",
                    flush=True,
                )

    log(
        f"[EDGAR scan] Processed {processed} filings → "
        f"{len(all_purchases)} qualifying purchases (>${MIN_PURCHASE_VALUE:,})",
        flush=True,
    )

    # Rank by dollar value descending
    all_purchases.sort(key=lambda x: x["total_value"], reverse=True)
    return all_purchases


# ---------------------------------------------------------------------------
# Formatting helpers
# ---------------------------------------------------------------------------

def fmt_money(v: float) -> str:
    if v >= 1_000_000:
        return f"${v / 1_000_000:.2f}M"
    return f"${v / 1_000:.0f}K"


def build_slack_message(purchases: list, start_date: str, end_date: str) -> str:
    now_str = datetime.now().strftime("%Y-%m-%d %H:%M UTC")

    if not purchases:
        return (
            f"*SEC EDGAR — Insider Purchases >$100K* | {start_date} → {end_date}\n"
            f"_No qualifying open-market purchases found in this window._\n"
            f"_Scanned at {now_str}_"
        )

    lines = [
        f"*:chart_with_upwards_trend: SEC Insider Buys — Top {min(len(purchases), 20)} Purchases >$100K*",
        f"_{start_date} → {end_date} | Scanned {now_str}_",
        "",
        "| # | Ticker | Company | Insider | Title | Shares | Price | Total |",
        "|---|--------|---------|---------|-------|--------|-------|-------|",
    ]

    for i, p in enumerate(purchases[:20], 1):
        ticker = f"`{p['ticker']}`" if p["ticker"] else "—"
        company = p["company"][:28] + ("…" if len(p["company"]) > 28 else "")
        insider = p["insider"][:22] + ("…" if len(p["insider"]) > 22 else "")
        title = p["title"][:18] + ("…" if len(p["title"]) > 18 else "")
        lines.append(
            f"| {i} | {ticker} | {company} | {insider} | {title} "
            f"| {p['shares']:,} | ${p['price']:.2f} | **{fmt_money(p['total_value'])}** |"
        )

    lines += [
        "",
        f"_Source: SEC EDGAR Form 4 filings · Filtered: open-market purchases (code P) · Threshold: $100K_",
    ]
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    purchases = run_scan(verbose=True)
    start_date, end_date = get_date_range(days=1)
    msg = build_slack_message(purchases, start_date, end_date)
    print("\n--- SLACK MESSAGE PREVIEW ---")
    print(msg)
    print(json.dumps(purchases[:5], indent=2))
